privilege/privileges/privileges_test.go

```diff
@@ -2084,3 +2084,55 @@ func TestGrantReferences(t *testing.T) {
 	tk.MustExec("DROP USER referencesUser")
 	tk.MustExec("DROP SCHEMA reftestdb")
 }
+
+// https://github.com/pingcap/tidb/issues/27213
+func TestShowGrantsWithRolesAndDynamicPrivs(t *testing.T) {
+	t.Parallel()
+
+	store, clean := newStore(t)
+	defer clean()
+
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("CREATE ROLE tsg_r1")
+	tk.MustExec("CREATE USER tsg_u1, tsg_u2")
+	tk.MustExec("GRANT CONNECTION_ADMIN, ROLE_ADMIN, SYSTEM_VARIABLES_ADMIN, PROCESS ON *.* TO tsg_r1")
+	tk.MustExec("GRANT CONNECTION_ADMIN ON *.* TO tsg_u1 WITH GRANT OPTION") // grant a superior privilege to the user
+	tk.MustExec("GRANT CONNECTION_ADMIN ON *.* TO tsg_u2 WITH GRANT OPTION") // grant a superior privilege to the user
+	tk.MustExec("GRANT ROLE_ADMIN ON *.* TO tsg_u1")
+	tk.MustExec("GRANT ROLE_ADMIN ON *.* TO tsg_u2")
+	tk.MustExec("GRANT ROLE_ADMIN ON *.* TO tsg_r1 WITH GRANT OPTION") // grant a superior privilege to the role
+
+	tk.MustExec("GRANT tsg_r1 TO tsg_u1, tsg_u2")    // grant the role to both users
+	tk.MustExec("SET DEFAULT ROLE tsg_r1 TO tsg_u1") // u1 has the role by default, but results should be identical.
+
+	// login as tsg_u1
+	tk.Session().Auth(&auth.UserIdentity{
+		Username: "tsg_u1",
+		Hostname: "localhost",
+	}, nil, nil)
+	tk.MustQuery("SHOW GRANTS").Check(testkit.Rows(
+		"GRANT PROCESS ON *.* TO 'tsg_u1'@'%'",
+		"GRANT 'tsg_r1'@'%' TO 'tsg_u1'@'%'",
+		"GRANT SYSTEM_VARIABLES_ADMIN ON *.* TO 'tsg_u1'@'%'",
+		"GRANT CONNECTION_ADMIN,ROLE_ADMIN ON *.* TO 'tsg_u1'@'%' WITH GRANT OPTION",
+	))
+
+	// login as tsg_u2 + SET ROLE
+	tk.Session().Auth(&auth.UserIdentity{
+		Username: "tsg_u2",
+		Hostname: "localhost",
+	}, nil, nil)
+	tk.MustQuery("SHOW GRANTS").Check(testkit.Rows(
+		"GRANT USAGE ON *.* TO 'tsg_u2'@'%'",
+		"GRANT 'tsg_r1'@'%' TO 'tsg_u2'@'%'",
+		"GRANT ROLE_ADMIN ON *.* TO 'tsg_u2'@'%'",
+		"GRANT CONNECTION_ADMIN ON *.* TO 'tsg_u2'@'%' WITH GRANT OPTION",
+	))
+	tk.MustExec("SET ROLE tsg_r1")
+	tk.MustQuery("SHOW GRANTS").Check(testkit.Rows(
```

Comment:
Also add a test for `show grants for current_user();`

Reply:
This should be fixed, PTAL again.

Reply:
Look like the following test case is still not fixed:

```sql
CREATE ROLE 'app_read';
create user test@'%';
GRANT SYSTEM_VARIABLES_ADMIN, PROCESS on *.* to 'app_read'@'%';
GRANT 'app_read' TO 'test'@'%';
SET DEFAULT ROLE app_read TO 'test'@'%';
```
2.
```sql
▶ mysql -u test -h 127.0.0.1 -P 4000
(none)> show grants for current_user();  -- the following result is not expected.
+------------------------------------+
| Grants for test@127.0.0.1          |
+------------------------------------+
| GRANT USAGE ON *.* TO 'test'@'%'   |
| GRANT 'app_read'@'%' TO 'test'@'%' |
+------------------------------------+
2 rows in set
Time: 0.009s
(none)> show grants;
+---------------------------------------------------+
| Grants for User                                   |
+---------------------------------------------------+
| GRANT PROCESS ON *.* TO 'test'@'%'                |
| GRANT 'app_read'@'%' TO 'test'@'%'                |
| GRANT SYSTEM_VARIABLES_ADMIN ON *.* TO 'test'@'%' |
+---------------------------------------------------+

```

Reply:
You're right! [This comment](https://github.com/pingcap/tidb/blob/dfe9acfa43237ceadab933420e4db20c6ad0c709/executor/builder.go#L745-L746) led me astray, but I've fixed it so that `SHOW GRANTS` and `SHOW GRANTS FOR CURRENT_USER()` should be the same now.